hw/formal/tools/jaspergold/parse-fpv-report.py

```diff
@@ -0,0 +1,180 @@
+    """
+    This extracts messages from the sting buffer full_file.
+    The argument patterns needs to be a list of tuples with
+    (<key>, <pattern_to_match_for>).
+    """
+    for severity, pattern in patterns:
+        results[severity] = re.findall(pattern, full_file, flags=re.MULTILINE)
+
+    return results
+
+
+def extract_messages_count(full_file, patterns, results):
+    """
+    This extracts how many times the messages appeared from the sting buffer full_file.
+    The argument patterns needs to be a list of tuples with
+    (<key>, <pattern_to_match_for>).
+    """
+    for severity, pattern in patterns:
+        results[severity] = len(re.findall(pattern, full_file, flags=re.MULTILINE))
+
+    return results
+
+
+def get_results(resdir):
+    """
+    Parse FPV log file and extract errors, warnings, and each property's verification status.
+    """
+    results = OrderedDict()
+    messages = OrderedDict()
+    fpv_summary = OrderedDict()
+    try:
+        with Path(resdir).joinpath('fpv.log').open() as f:
+            full_file = f.read()
+            err_warn_patterns = [("errors", r"^ERROR: .*"),
+                                 ("warnings", r"^WARNING: .*"),
+                                 ("fpv_errors", r"^\[ERROR.*"),
+                                 ("fpv_warnings", r"^\[WARN.*"),
+                                 ("cex", r"^\[\d+\]\s+(\S*)\s+cex.*"),
+                                 ("undetermined", r"^\[\d+\]\s+(\S*)\s+undetermined.*"),
+                                 ("unreachable", r"^\[\d+\]\s+(\S*)\s+unreachable.*")]
+            extract_messages(full_file, err_warn_patterns, messages)
+
+            message_patterns = [("fpv_errors", r"^\[ERROR.*"),
+                                ("fpv_warnings", r"^\[WARN.*"),
+                                ("proven", r"^\[\d+\].*proven.*"),
+                                ("cex", r"^\[\d+\].*cex.*"),
+                                ("covered", r"^\[\d+\].*covered.*"),
+                                ("undetermined", r"^\[\d+\].*undetermined.*"),
+                                ("unreachable", r"^\[\d+\].*unreachable.*")]
+            extract_messages_count(full_file, message_patterns, fpv_summary)
+
+            total_msg = fpv_summary["proven"] + fpv_summary["cex"] + fpv_summary["undetermined"]
+            if total_msg > 0:
+                pass_rate = (fpv_summary["proven"] + fpv_summary["undetermined"]) / total_msg * 100
+            else:
+                pass_rate = 0
+            fpv_summary["pass_rate"] = "{0:.2f} %".format(round(pass_rate, 2))
+
+            total_msg = fpv_summary["covered"] + fpv_summary["unreachable"]
+            if total_msg > 0:
+                cov_rate = fpv_summary["covered"] / total_msg * 100
+            else:
+                cov_rate = 0
+            fpv_summary["cov_rate"] = "{0:.2f} %".format(round(cov_rate, 2))
+
+    except IOError as err:
+        messages["errors"] += ["IOError: %s" % err]
```

Comment:
Nit: I'd probably write this as a call to append: `messages['errors'].append('IOError: %s' % err)`. Doesn't really matter, but it seems less nice to construct a singleton list then add it on to the end of another list when there's a function that will do it directly.